pkg/aws/policy_document.go

```diff
@@ -121,6 +121,81 @@ func (p *PolicyDocument) IsActionAllowed(wanted string) bool {
 	return false
 }
 
+func (p *PolicyDocument) GetAllowedActions() []string {
+	var actions []string
+	for _, statement := range p.Statement {
+		if statement.Effect != "Allow" {
+			continue
+		}
+		switch action := statement.Action.(type) {
+		case string:
+			actions = append(actions, action)
+		case []interface{}:
+			for _, el := range action {
+				actions = append(actions, el.(string))
+			}
+		}
+	}
+	return actions
+}
+
+// checkPermissionsUsingQueryClient will use queryClient to query whether the credentials in targetClient can perform
+// the actions listed in the statementEntries. queryClient will need
+// sts:GetCallerIdentity and iam:SimulatePrincipalPolicy
+func (p *PolicyDocument) checkPermissionsUsingQueryClient(queryClient *awsClient, targetUserARN string,
+	params *SimulateParams) (bool, error) {
+	// Ignoring isRoot here since we only warn the user that its not best practice to use it.
+	// TODO: Add a check for isRoot in the initialize
+	allowList := aws.StringSlice(p.GetAllowedActions())
+
+	input := &iam.SimulatePrincipalPolicyInput{
+		PolicySourceArn: aws.String(targetUserARN),
+		ActionNames:     allowList,
+		ContextEntries:  []*iam.ContextEntry{},
+	}
+
+	if params != nil {
+		if params.Region != "" {
+			input.ContextEntries = append(input.ContextEntries, &iam.ContextEntry{
+				ContextKeyName:   aws.String("aws:RequestedRegion"),
+				ContextKeyType:   aws.String("stringList"),
+				ContextKeyValues: []*string{aws.String(params.Region)},
+			})
+		}
+	}
+
+	// Collect all failed actions
+	var failedActions []string
+	err := queryClient.iamClient.SimulatePrincipalPolicyPages(input,
+		func(response *iam.SimulatePolicyResponse, lastPage bool) bool {
+			for _, result := range response.EvaluationResults {
+				if *result.EvalDecision != "allowed" {
+					// Don't bail out after the first failure, so we can log the full list
+					// of failed/denied actions
+					failedActions = append(failedActions, *result.EvalActionName)
+				}
+			}
+			return !lastPage
+		})
+	if err != nil {
+		return false, fmt.Errorf("Error simulating policy: %v", err)
+	}
+
+	if len(failedActions) > 0 {
+		return false, fmt.Errorf("Actions not allowed with tested credentials: %v", failedActions)
+	}
+
+	return true, nil
+}
+
+func (p PolicyDocument) String() string {
```

Comment:
This is not a pointer receiver on purpose - if you pass a `PolicyDocument` or `*PolicyDocument` to any formatted printer, this will print it properly while if this was a pointer receiver it would only print properly when passed a pointer.